### PR TITLE
optimize pyreader

### PR DESCRIPTION
**Pull request by jacquesqiao**

### CODE REVIEW

File: paddle/fluid/CMakeLists.txt [hunk 1]
```diff
@@ -12,6 +12,5 @@ endif(NOT WIN32)
 if(WITH_INFERENCE)
   # NOTE: please add subdirectory inference at last.
   add_subdirectory(inference)
+  add_subdirectory(train)
```
Comment (panyx0718) [inline paddle/fluid/CMakeLists.txt:15]: why? seems OK to test without inference?

File: python/paddle/fluid/layers/io.py [hunk 2]
```diff
@@ -470,6 +471,153 @@ def random_data_generator(low, high, shapes, lod_levels, for_parallel=True):
...
 
+def _py_reader(capacity,
+               shapes,
+               dtypes,
+               lod_levels=None,
+               name=None,
+               use_double_buffer=True,
+               feed_list=None):
+    if feed_list is not None:
+        assert isinstance(feed_list, list)
```
Comment (panyx0718) [inline python/paddle/fluid/layers/io.py]: some user-friendly message?

File: python/paddle/fluid/tests/unittests/test_py_reader_using_executor.py [hunk 3]
```diff
@@ -53,13 +53,22 @@ def simple_fc_net(in_size,
...
-        use_double_buffer=False)
+                  use_double_buffer=False,
+                  use_feed_list=True):
+    if use_feed_list:
+        data = fluid.layers.data(name="data", dtype='float32', shape=[in_size])
+        label = fluid.layers.data(name='label', dtype='int64', shape=[1])
+        reader = fluid.layers.py_reader_by_data(
+            capacity=queue_capacity,
+            use_double_buffer=False,
+            feed_list=[data, label])
```
Comment (panyx0718) [inline python/paddle/fluid/tests/unittests/test_py_reader_using_executor.py:64]: Is there a test to check that user can feed data with py_reader?

File: python/paddle/fluid/layers/io.py [hunk 4]
```diff
@@ -596,128 +744,28 @@ def py_reader(capacity,
...
-    reader.reset = __reset__
-    reader.decorate_tensor_provider = __set_tensor_provider__
-    reader.decorate_paddle_reader = __set_paddle_reader__
-    reader.start = __start__
-
-    return reader
+def py_reader_by_data(capacity, feed_list, name=None, use_double_buffer=True):
+    """
+    Works much like py_reader except that it's input is feed_list
+    instead of shapes, dtypes, lod_levels
```
Comment (panyx0718) [inline python/paddle/fluid/layers/io.py]: formal comment?

File: python/paddle/fluid/layers/io.py [hunk 5]
```diff
@@ -470,6 +471,153 @@ def random_data_generator(low, high, shapes, lod_levels, for_parallel=True):
...
+        shape_concat = []
+        ranks = []
+        shapes = []
+
+        for data in feed_list:
+            dtypes.append(data.dtype)
+            shape_concat.extend(data.shape)
+            ranks.append(len(data.shape))
+            shapes.append(data.shape)
+            lod_levels.append(data.lod_level)
```
Comment (panyx0718) [inline python/paddle/fluid/layers/io.py:497]: will names come from data?

Comment (jacquesqiao) [reply python/paddle/fluid/layers/io.py:497]: yes, the variable for data is created by data list

File: paddle/fluid/CMakeLists.txt (hunk 1 above)
Comment (jacquesqiao) [reply paddle/fluid/CMakeLists.txt:15]: `paddle_fluid_origin` 是定义在inference中的
https://github.com/PaddlePaddle/Paddle/blob/develop/paddle/fluid/inference/CMakeLists.txt#L18

File: python/paddle/fluid/tests/unittests/test_py_reader_using_executor.py (hunk 3 above)
Comment (jacquesqiao) [reply python/paddle/fluid/tests/unittests/test_py_reader_using_executor.py:64]: This is the test for use_feed_list.

Comment (panyx0718) [reply python/paddle/fluid/tests/unittests/test_py_reader_using_executor.py:64]: I mean during executor.run, can you feed?

Comment (jacquesqiao) [reply python/paddle/fluid/tests/unittests/test_py_reader_using_executor.py:64]: No, it's another mode.

File: python/paddle/fluid/layers/io.py (hunk 4 above)
Comment (jacquesqiao) [reply python/paddle/fluid/layers/io.py]: done

**Review by panyx0718 — DISMISSED** [review]

File: python/paddle/fluid/layers/io.py [hunk 6]
```diff
@@ -594,128 +747,72 @@ def py_reader(capacity,
...
-    reader.start = __start__
+        1. The basic usage of :code:`py_reader` is as follows:
 
-    return reader
+        >>> import paddle.fluid as fluid
+        >>> import paddle.dataset.mnist as mnist
+        >>>
+        >>> image = fluid.layers.data(name='image', shape=[3,224,224], dtypes='float32')
+        >>> label = fluid.layers.data(name='label', shape=[1], dtypes='int64')
+        >>> reader = fluid.layers.py_reader(capacity=64, feed_list=[image, label])
```
Comment (typhoonzero) [inline python/paddle/fluid/layers/io.py]: `fluid.layers. create_py_reader_by_data(...)`

File: python/paddle/fluid/layers/io.py [hunk 7]
```diff
@@ -594,128 +747,72 @@ def py_reader(capacity,
...
+        >>> loss = network(img, label) # some network definition
+        >>>
+        >>> fluid.Executor(fluid.CUDAPlace(0)).run(fluid.default_startup_program())
+        >>>
+        >>> exe = fluid.ParallelExecutor(use_cuda=True, loss_name=loss.name)
+        >>> for epoch_id in range(10):
+        >>>     reader.start()
+        >>>     try:
+        >>>         while True:
+        >>>             exe.run(fetch_list=[loss.name])
```
Comment (typhoonzero) [inline python/paddle/fluid/layers/io.py:804]: can `exe.run` auto handle EOF exception? or it's in another PR?

Comment (jacquesqiao) [reply python/paddle/fluid/layers/io.py:804]: Will add in following PR

File: python/paddle/fluid/layers/io.py (hunk 6 above)
Comment (jacquesqiao) [reply python/paddle/fluid/layers/io.py]: done

**Review by typhoonzero — APPROVED** [review]
LGTM

**Review by panyx0718 — APPROVED** [review]

**Pull request merged**